modules/tinymce/src/themes/silver/main/ts/ui/dialog/ImagePreview.ts

```diff
@@ -0,0 +1,164 @@
+    dom: {
+      tag: 'img',
+      classes: [ 'tox-imagepreview__image' ],
+      attributes: initialData.map((data) => ({ src: data.url })).getOr({})
+    },
+  });
+
+  const memContainer = Memento.record({
+    dom: {
+      tag: 'div',
+      classes: [ 'tox-imagepreview__container' ],
+      attributes: {
+        role: 'presentation'
+      },
+    },
+    components: [
+      memImage.asSpec()
+    ]
+  });
+
+  const setValue = (frameComponent: AlloyComponent, data: ImagePreviewDataSpec) => {
+    const translatedData: Dialog.ImagePreviewData = {
+      url: data.url
+    };
+    // update properties that are set by the data
+    data.zoom.each((z) => translatedData.zoom = z);
+    data.cachedWidth.each((z) => translatedData.cachedWidth = z);
+    data.cachedHeight.each((z) => translatedData.cachedHeight = z);
+    cachedData.set(translatedData);
+
+    const applyFramePositioning = () => {
+      const imageWidth = translatedData.cachedWidth;
+      const imageHeight = translatedData.cachedHeight;
+
+      if (translatedData.zoom === undefined) {
+        const z = zoomToFit(frameComponent.element, imageWidth, imageHeight);
+        // sneaky mutation since we own the object
+        translatedData.zoom = z;
+      }
+      const position = calculateImagePosition(
+        Width.get(frameComponent.element),
+        Height.get(frameComponent.element),
+        imageWidth, imageHeight,
+        translatedData.zoom
+      );
+      memContainer.getOpt(frameComponent).each((container) => {
+        Css.setAll(container.element, position);
+      });
+    };
+
+    memImage.getOpt(frameComponent).each((imageComponent) => {
+      const img = imageComponent.element;
+      if (data.url !== Attribute.get(img, 'src')) {
+        Attribute.set(img, 'src', data.url);
+        Class.remove(frameComponent.element, 'tox-imagepreview__loaded');
+      }
+
+      if (translatedData.cachedWidth !== undefined && translatedData.cachedHeight !== undefined) {
+        applyFramePositioning();
+      }
+
+      Ready.image(img).then((img) => {
+        // Ensure the component hasn't been removed while the image was loading
+        // if it is disconnected, just do nothing
+        if (frameComponent.getSystem().isConnected()) {
+          Class.add(frameComponent.element, 'tox-imagepreview__loaded');
+          // sneaky mutation since we own the object
+          translatedData.cachedWidth = img.dom.naturalWidth;
+          translatedData.cachedHeight = img.dom.naturalHeight;
+          applyFramePositioning();
+        }
+      });
+    });
+  };
+
+  const styles: Record<string, string> = {};
+  spec.height.each((h) => styles.height = h);
+
+  // TODO: Use the initial data properly once it's validated
+  const fakeValidatedData: Optional<ImagePreviewDataSpec> = initialData.map((d) => ({
+    url: d.url,
+    zoom: Optional.from(d.zoom),
+    cachedWidth: Optional.from(d.cachedWidth),
+    cachedHeight: Optional.from(d.cachedHeight),
+  }));
```

Comment:
This is a known issue due to the way `initialData` is not validated but `setData` calls are. We decided to leave it in place and hope to address it before release.

Reply:
```suggestion
  // TODO: TINY-8393 Use the initial data properly once it's validated
  const fakeValidatedData: Optional<ImagePreviewDataSpec> = initialData.map((d) => ({
    url: d.url,
    zoom: Optional.from(d.zoom),
    cachedWidth: Optional.from(d.cachedWidth),
    cachedHeight: Optional.from(d.cachedHeight),
  }));
```